packages/block-library/src/separator/edit.js

```diff
@@ -6,29 +6,57 @@ import classnames from 'classnames';
 /**
  * WordPress dependencies
  */
-import { HorizontalRule } from '@wordpress/components';
+import {
+	HorizontalRule,
+	__experimentalBoxControl as BoxControl,
+} from '@wordpress/components';
+import { View } from '@wordpress/primitives';
 import { withColors, useBlockProps } from '@wordpress/block-editor';
+
 /**
  * Internal dependencies
  */
 import SeparatorSettings from './separator-settings';
 
-function SeparatorEdit( { color, setColor, className } ) {
+const { __Visualizer: BoxControlVisualizer } = BoxControl;
+
+function SeparatorEdit( props ) {
+	const {
+		color,
+		attributes: { style },
+	} = props;
+
+	const { top, bottom } = style?.spacing?.margin || {};
+	const blockProps = useBlockProps();
+
 	return (
 		<>
-			<HorizontalRule
-				{ ...useBlockProps( {
-					className: classnames( className, {
+			<View
+				{ ...blockProps }
+				className={ blockProps.className?.replace(
+					'wp-block-separator',
+					'wp-block-separator-wrapper'
+				) }
+				style={ {} }
+			>
+				<BoxControlVisualizer
+					values={ style?.spacing?.margin }
+					showValues={ style?.visualizers?.margin }
+				/>
+				<HorizontalRule
+					className={ classnames( blockProps.className, {
 						'has-background': color.color,
 						[ color.class ]: color.class,
-					} ),
-					style: {
+					} ) }
+					style={ {
 						backgroundColor: color.color,
 						color: color.color,
-					},
-				} ) }
-			/>
-			<SeparatorSettings color={ color } setColor={ setColor } />
+						marginTop: top,
+						marginBottom: bottom,
+					} }
+				/>
+			</View>
+			<SeparatorSettings { ...props } />
```

Comment:
Any reason we need all these changes to the block's edit function. The idea of the block supports is to "automatically" add support without changes, what's preventing us from doing that?

Reply:
The `hr` element used for the separator presents a couple of problems in the editor, mainly the ability to select it and visualize the changes made to margins.

The original issues related to this block expressed a desire to control the "height" which really related to margins instead. If we don't have a true height for the separator it makes the separator rather hard to select by click on it with the mouse. Adding a wrapper around the `hr`  allows the margins to expand within a clickable container drastically improving selection.

Earlier iterations on this included a resizable box so the user could drag to resize the separator. While it was decided the drag resizing overcomplicated the block, keeping the visualizer was still desirable. The visualizer couldn't be added within the `hr` so also benefits from being in the new wrapper element.

The additional changes to the application of the styles were to make the separator display as expected.

In terms of saving the margins, that works as per normal "automatic" block support. The frontend is still only a simple `hr` with top/bottom margins. The bulk of the other changes in this PR are to offer the same functionality in native for the separator block as it has on the web.

Reply:
I see the challenges here, though the extra wrapper creates a diversion between the frontend and backend that we tried to illiminate in blocks over time because it makes styling the editor for themes very hard and create a lot of inconsistencies.

It's maybe fine for this particular block but would be good to have themers perspective. cc @scruffian @carolinan 

> The visualizer couldn't be added within the hr so also benefits from being in the new wrapper element.

I'm aware of this and it's the reason we removed the visualizer temporarily from other blocks like the "group" block. I think it's be good to have the visualizer work a bit differently to solve this (don't rely on a parent element but maybe a "ref"/popover)

Do you think these issues: visualizer and block selection are absolutely necessary in this PR or do you think we can iterate on them separately while still keeping the front-backend parity (maybe using alternative solutions like the proposal above)?

Reply:
> Do you think these issues: visualizer and block selection are absolutely necessary in this PR or do you think we can iterate on them separately while still keeping the front-backend parity

The visualizer and easier block selection are not non-negotiable. 

I've updated this PR to remove the visualizer and minimise changes. The majority of the updates are now related to Native and adding color and margin controls there. The PR has been rebased as well.